test/EndToEndSpec.hs

```diff
@@ -469,6 +469,69 @@ spec = around_ (inTempDirectoryNamed "foo") $ do
           - "*.markdown"
         |] `shouldWarn` ["Specified pattern \"*.markdown\" for extra-doc-files does not match any files"]
 
+    describe "build-tools" $ do
+      it "adds known build tools to build-tools" $ do
+        [i|
+        executable:
+          build-tools:
+            alex == 0.1.0
+        |] `shouldRenderTo` executable_ "foo" [i|
+        build-tools:
+            alex ==0.1.0
+        |]
+
+      it "adds other build tools to build-tool-depends" $ do
+        [i|
+        executable:
+          build-tools:
+            hspec-discover: 0.1.0
+        |] `shouldRenderTo` (executable_ "foo" [i|
+        build-tool-depends:
+            hspec-discover:hspec-discover ==0.1.0
+        |]) {
+          -- NOTE: We do not set this to 2.0 on purpose, so that the .cabal
+          -- file is compatible with a wider range of Cabal versions!
```

Comment:
What's the meaning of 'compatible' here? Will the `.cabal` file actually work as desired with old Cabals, or will the file be read without complaints but the tool not be run? If it's the latter, I think we should put a proper constraint on.

Reply:
As per https://cabal.readthedocs.io/en/latest/developing-packages.html#buildtoolsbc, for older versions of `cabal-install` a user has to make sure that required build tools are on the `PATH`.

The specific case of `hspec-discover` will work with older versions of `cabal-install` without any additional user effort, as `hspec-discover` is installed as a dependency of `hspec`.

Reply:
OK; I think that that definitely needs a mention in the docs. `cabal check` should also be complaining about this because `build-tool-depends` is new in 2.0, but I see that it actually isn't, but we can't rely on that.

Reply:
> OK; I think that that definitely needs a mention in the docs. 

What exactly do you think should we add to the docs?

> `cabal check` should also be complaining about this because `build-tool-depends` is new in 2.0, but I see that it actually isn't, but we can't rely on that.

I think this is by design for the purpose of backwards compatibility.  This is at least how I read https://cabal.readthedocs.io/en/latest/developing-packages.html#buildtoolsbc.

Reply:
https://cabal.readthedocs.io/en/latest/developing-packages.html#pkg-field-build-tool-depends says that it's ignored with a warning on old versions of Cabal.

Suggested docs warning:

> Note that packages using `build-tool-depends` will likely not work correctly with versions of Cabal prior to 2.0, because the build tools will not be run.

This is why I was asking about what counted as 'compatibility': packages that wind up using `build-tool-depends` will likely break when they're built on old Cabals, unless they're using the build tools for something non-essential; in a real sense they're not compatible with old Cabals, even if those old Cabals can read the rest of the package description. ISTM we'd be better off giving those `cabal-version: 2.0` and dodging both future warnings and breakage.

Reply:
> packages that wind up using `build-tool-depends` will likely break when they're built on old Cabals

The large majority of packages that use `build-tool-depends` use `hspec-discover`.  At least these packages will continue to work with old versions of `cabal-install`.  The same is true for packages that use `markdown-unlit`.

For other packages that use `build-tool-depends`, they may work as well:

1. either because the executable is pulled in as a dependency (as is the case for `hspec-discover`)
1. or because the executable is on the `PATH` by other means

So for many packages there is a good chance that a package that uses `build-tool-depends` will work with an old version of `cabal`.

If it does not work, `cabal` will fail due to the missing executable at build time.  If we would require `cabal-version: 2.0` it would fail at configuration time.

Failing at configuration time is strictly better than failing at build time. But the question is still which of the following options do we prefer:

1. Have the majority of packages that use `build-tool-depends` work with old versions of `cabal` and the ones that don't work fail with a build error
1. Have all packages that use `build-tool-depends` fail at configuration time with old versions of `cabal`

> ISTM we'd be better off giving those `cabal-version: 2.0` and dodging both future warnings and breakage.

I don't expect there to be any future warnings.  As I understand it `build-tool-depends` was carefully implemented to allow for this kind of backwards compatibility, arguably at the cost of comprehensibility.